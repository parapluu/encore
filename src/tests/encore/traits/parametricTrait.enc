trait T
def id<a>(x : a) : a
  x
def phantom<a, b>(x: a): Maybe b
  Nothing

passive class Foo<s> : T

class Main
  def main() : void
    let f = (new Foo<int> : T).id<int>(42)
        f' = (new Foo<int> : T).phantom<int, String>(42)
    in {
      print("Id result: {}\n", f);
      match f' with
      Nothing => print("Phantom type is 'Nothing'")
      Just s => print("ERROR")
    }
