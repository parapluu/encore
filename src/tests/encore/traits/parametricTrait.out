Id result: 42
Phantom type is 'Nothing'
